### ISSUE TITLE
Leo-Bridge can be wrong about an outline read from a file

### ISSUE THREAD

**Issue by SegundoBob** [leo-editor/leo-editor#1437]
## Description
Before reading a file, Leo generates one or more nodes with GNX's for the file.  Usually this is harmless because the GNX's generated do not match any of the GNX's in the outline being read.  But when a generated GNX equals a GNX in the outline being read, Leo is misled about the outline contents.

Issue [#1348](https://github.com/leo-editor/leo-editor/issues/1348) describes how to cause a GNX conflict with great frequency.  This requires using Leo-Bridge in one process to create a node.  Then within a second of the node creation, using another process to open the file containing the just created node.  This problem has occurred during valid, everyday, non-contrived use of Leo-Bridge.

## Possible Solutions

This Leo-Editor Forum thread [Bob H: would distinct Leo IDs help with #1348?](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/leo-editor/My2DL1POr5I/akYyGnTEBAAJ) suggests two possible fixes:

1) Do not generate any GNX that might affect the handling of the file being read until after the file is read and all of its GNX's are known.

2) If GNX's are generated before a file is read, then delete these GNX's before they can conflict with GNX's in the file.

**Comment by edreamleo** [comment]
@SegundoBob Thanks for this concise summary. I've marked it for early review.

**Comment by SegundoBob** [comment]
## Another Possible Solution

If and only if the Leo-Editor initiated nodes are never written to a file, then GNX conflicts can easily be avoided in the same way that conflicts with the hidden root node GNX is avoided.  That is, the Leo-Editor initiated nodes can be given GNX's that don't match any ordinary GNX.

Example GNX's for Leo-Editor initiated nodes:

*** leo-internal.1 ***
*** leo-internal.2 ***
etc.

**Comment by btheado** [comment]
According to git blame, the 6a226718d commit seems to be the one which added the early call to leo/core/leoFrame.py(706)createFirstTreeNode() (see https://github.com/leo-editor/leo-editor/issues/1348#issuecomment-542014635 for full call stack).  It is this early call which is causing the GNX collision that Bob is seeing. ("# 2017/11/12: For #503: Use string/null gui for unit tests.").

I just stumbled upon this comment in the code: https://github.com/leo-editor/leo-editor/blob/master/leo/core/leoBridge.py#L277 which led me to issues #130 and #35. Those issues are somewhat similar to this one since they also involve the leo file being opened and closed in the same second. There is lots of discussion there, so I thought I'd drop the links here for reference.

**Comment by edreamleo** [comment]
**The fix**

LeoFrame.createFirstTreeNode now *always* sets c.hiddenRootNode.gnx to 'hidden-root-vnode-gnx'.

That's *all*. This gnx *never* changes, and can never match a "real" gnx. Job done.


The fix was unexpected easy.

**Comment by vitalije** [comment]
@edreamleo  the proposed fix introduces potentially dangerous condition in the outline. Two different vnodes have the same gnx and yet one is parent of the other. You can see it if you add the following two lines at the end of createFirstTreeNode:

```python
def createFirstTreeNode(self):
     ...
     print(c.hiddenRootNode.gnx == c.hiddenRootNode.children[0].gnx)
     print(c.hiddenRootNode is v)
```
If it fixes the bug, then it would be better not to use `'hidden-root-vnode-gnx'` as gnx but some other constant like `'empty-outline-first-vnode-gnx'`.
The effect I believe would be the same but it won't have this gnx collision condition.

**Comment by btheado** [comment]
Here is a simplified pytest which duplicates the issue in current `devel` branch:

```python
import leo.core.leoBridge as leoBridge
import pytest
def bridge():
    return leoBridge.controller(gui='nullGui',
        loadPlugins=False, readSettings=False,
        silent=False, verbose=False, useCaches=False)

def test_1437(tmp_path):
    # Create, save and close outline with a single node
    fileName = str(tmp_path.joinpath("tree1.leo"))
    c = bridge().openLeoFile(fileName)
    p = c.rootPosition()
    p.h = "1"
    assert c.rootPosition().h == "1"
    c.save()
    c.close()

    # Reset the global state so a new bridge will come
    # with fresh nodeIndices. This simulates
    # SegundoBob's launching bridge consecutively in
    # multiple processes
    leoBridge.gBridgeController = None

    # Read the file and verify the headline is "1"
    c = bridge().openLeoFile(fileName)
    assert c.rootPosition().h == "1"
```

The failure output looks like this (Note it is conceivable the test won't fail every time since the above code could straddle two different seconds of time. It should run fast enough to not be likely, but it is entirely possible):

```
        c = bridge().openLeoFile(fileName)
>       assert c.rootPosition().h == "1"
E       AssertionError: assert 'NewHeadline' == '1'
E         - NewHeadline
E         + 1

leo/test/pytest/1437_test.py:31: AssertionError
----------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------
setting leoID from os.getenv('USER'): 'btheado'
file not found: /tmp/pytest-of-btheado/pytest-46/test_14370/tree1.leo. creating new window
setting leoID from os.getenv('USER'): 'btheado'
```

Even in the `gnx` branch the test doesn't pass (maybe this is what @vitalije is concerned about):

```
        c = bridge().openLeoFile(fileName)
>       assert c.rootPosition().h == "1"
E       AttributeError: 'NoneType' object has no attribute 'h'

leo/test/pytest/1437_test.py:31: AttributeError
----------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------
setting leoID from os.getenv('USER'): 'btheado'
file not found: /tmp/pytest-of-btheado/pytest-47/test_14370/tree1.leo. creating new window
setting leoID from os.getenv('USER'): 'btheado'
vnode: <VNode hidden-root-vnode-gnx <hidden root vnode>> is its own parent
check-outline ERROR! tree1.leo 1 nodes, 0 gnx errors, 1 structure errors
vnode: <VNode hidden-root-vnode-gnx <hidden root vnode>> is its own parent
check-outline ERROR! tree1.leo 1 nodes, 0 gnx errors, 1 structure errors
setCurrentPosition Invalid position: '<hidden root vnode>'
setCurrentPosition finishOpen,outerUpdate,redraw,selectPosition
```

**Comment by SegundoBob** [comment]
Kudos, Brian.  Great work.

**Comment by SegundoBob** [comment]
Kudos, Vitaije.  Looks like your code analysis is correct.

SegundoBob

**Comment by edreamleo** [comment]
On Sat, Mar 7, 2020 at 10:26 AM Виталије Милошевић <notifications@github.com>
wrote:

> @edreamleo <https://github.com/edreamleo> the proposed fix introduces
> potentially dangerous condition in the outline. Two different vnodes have
> the same gnx and yet one is parent of the other.
>
Thanks for this. I was confused about what createFirstTreeNode.  It
allocates a *child *of c.hiddenRootNode, not c.hiddenRootNode itself. In
fact, c.hiddenRootNode (a vnode) already exists when
frame.createFirstTreeNode is called.

Here is the real complication. createFirstTreeNode creates a node which *might
*soon be deleted, or not. If the new root *isn't* deleted, then it must be
given a unique gnx *after* reading the file. I agree that the newly created
node should be given a gnx that conflicts neither with c.hiddenRootNode.gnx
nor with any gnx in the file.

In short, my first try was closer to the mark, but createFirstTreeNode
should, within the bridge, assign a second pseudo gnx. I'll give this a try
later today.

Edward

**Comment by edreamleo** [comment]
@vitalije Many thanks for your fix.

I am going to close this issue, and delete the gnx branch without merging it into devel.

**Comment by SegundoBob** [comment]
Vitalije,

Thanks for your fix.

Without your fix, my test detects an error about 70% of the time.
With your fix, my test never reports an error detection.

SegundoBob
```
Version:
commit 380024b2e8b51e3a510a0a4c9510d6f23ac8009f
Author: Edward K. Ream <edreamleo@gmail.com>
Date:   Sat Mar 7 12:19:24 2020 -0600

Test:
ic| test03.py:78 in main()- loopCnt: 27
^C
27 loops 19 errors 70% errors/loops

Version:
commit 55bd1556208e1d67b6b878260fbf0107e71f0d14 (HEAD -> devel, upstream/devel, origin/devel, origin/HEAD)
Merge: f772c5007 a57a57996
Author: Edward K. Ream <edreamleo@gmail.com>
Date:   Sun Mar 8 15:13:06 2020 -0500

Test:
ic| test03.py:78 in main()- loopCnt: 100
^C
100 loops 0 errors 0% errors/loops
```

**Comment by edreamleo** [comment]
On Sun, Mar 8, 2020 at 4:47 PM SegundoBob <notifications@github.com> wrote:

> Vitalije,
>
> Thanks for your fix.
>
> Without your fix, my test detects an error about 70% of the time.
> With your fix, my test never reports an error detection.
>
Hurray! It's been quite a saga. It's great that the fix is one line of code.

Edward
